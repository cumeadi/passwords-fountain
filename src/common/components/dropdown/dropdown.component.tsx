import { h, VNode } from 'preact';
import { TypedComponent } from '@/common/typings/prop-types';
import { useState } from 'preact/hooks';
import PropTypes from 'prop-types';
import {
    Wrapper,
    OptionsWrapper,
    Options,
    Option,
    SelectedOptionWrapper,
    ExpansionIcon,
} from './dropdown.styles';
import { renderIfTrue } from '@/common/utils/rendering';

export interface DropdownOption {
    id: string | number | symbol;
    label: string;
    value: string | number | symbol | object | Array<any> | null;
}

interface InitialDropdownOption extends DropdownOption {
    value: symbol;
}

const initialOption: InitialDropdownOption = {
    id: 'initial',
    label: 'dropdown.choose',
    value: Symbol('PLACEHOLDER_VALUE'),
};

export const Dropdown: TypedComponent<Props> = ({
    renderSelectedOption,
    renderOptionList,
    renderExpandIcon,
    withDefaultOption,
}: Props) => {
    const [expanded, setExpansion] = useState(false);
    const [selectedOption, setSelectedOption] = useState<DropdownOption>(
        withDefaultOption ?? initialOption
    );

    const onOptionClick = (option: DropdownOption): void => {
        setSelectedOption(option);
        setExpansion(false);
    };

    const renderOptions = () => {
        return renderIfTrue(() => (
            <Options>
                {renderOptionList(
                    Option,
                    onOptionClick,
                    selectedOption,
                    initialOption
                )}
            </Options>
        ))(expanded);
    };

    const onOptionSelection = (): void => {
        setExpansion(prevState => !prevState);
    };

    const onOutClick = (): void => {
        expanded && onOptionSelection();
    };

    const renderExpansionIcon = (): VNode => {
        if (renderExpandIcon) {
            return renderExpandIcon();
        }

        return <ExpansionIcon />;
    };

    return (
        <Wrapper tabindex={0} onblur={onOutClick}>
            <SelectedOptionWrapper onClick={onOptionSelection}>
                {renderSelectedOption(selectedOption, initialOption)}
                {renderExpansionIcon()}
            </SelectedOptionWrapper>
            <OptionsWrapper>{renderOptions()}</OptionsWrapper>
        </Wrapper>
    );
};

interface Props {
    renderOptionList: (
        OptionComponent: any,
        onOptionClick: (...args: any[]) => void,
        selectedOption: DropdownOption,
        initialOption: InitialDropdownOption
    ) => VNode[];
    renderSelectedOption: (
        selectedOption: DropdownOption,
        initialOption: DropdownOption
    ) => VNode | null;
    renderExpandIcon?: () => VNode;
    withDefaultOption?: DropdownOption;
}

Dropdown.propTypes = {
    renderOptionList: PropTypes.func.isRequired,
    renderSelectedOption: PropTypes.func.isRequired,
    renderExpandIcon: PropTypes.func,
    withDefaultOption: PropTypes.any,
};
