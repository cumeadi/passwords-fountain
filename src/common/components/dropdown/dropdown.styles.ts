import { h } from 'preact';
import { styled, setPragma } from 'goober';
import { theme, typography, stylelint } from '@/common/theme';
setPragma(h);

export const Wrapper = styled('section')`
    display: grid;
    min-width: 120px;
    max-width: 360px;
    text-transform: uppercase;

    &:focus {
        border: none;
        outline: none;
    }
`;

const optionCommon = stylelint.css`
    cursor: pointer;
    padding: ${theme.spacing.m18} ${theme.spacing.xl30};
`;

export const SelectedOptionWrapper = styled('div')`
    ${optionCommon}
    display: grid;
    grid-template-columns: 1fr auto;
    align-items: center;
    color: ${theme.colors.white};
    font-size: ${theme.fontSizes.s18};
    background: ${theme.colors.primaryBlue};

    &:hover {
        cursor: pointer;
        text-decoration: underline;
    }
`;

export const OptionListWrapper = styled('div')``;

export const OptionList = styled('ul')`
    border: 1px solid ${theme.colors.primaryBlue};
`;

export const Option = styled('li')`
    ${optionCommon}
    color: ${theme.colors.darkBlue};

    &:hover {
        cursor: pointer;
        text-decoration: underline;
        background: ${theme.colors.pastelGreen};
    }
`;

export const ExpansionIcon = styled('span')`
    width: 0;
    height: 0;
    border-left: 8px solid transparent;
    border-right: 8px solid transparent;
    border-top: 8px solid ${theme.colors.white};
`;
