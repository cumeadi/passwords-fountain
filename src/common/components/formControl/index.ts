export { FormControl } from './formControl.component';
