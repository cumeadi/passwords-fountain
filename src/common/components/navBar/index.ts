export { NavBar } from './navBar.component';
