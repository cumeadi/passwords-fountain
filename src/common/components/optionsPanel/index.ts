export { OptionsPanel } from './optionsPanel.component';
