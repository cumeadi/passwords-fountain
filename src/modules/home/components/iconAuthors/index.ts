export { IconAuthors } from './iconAuthors.component';
