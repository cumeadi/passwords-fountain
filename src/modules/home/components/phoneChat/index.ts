export { PhoneChat } from './phoneChat.component';
