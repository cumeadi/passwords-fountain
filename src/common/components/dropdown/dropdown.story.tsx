import { h, VNode } from 'preact';
import { Dropdown } from './dropdown.component';
import { withA11y } from '@storybook/addon-a11y';

export default {
    title: 'Dropdown',
    decorators: [withA11y],
};

export const defaultView = (): VNode => <Dropdown />;
