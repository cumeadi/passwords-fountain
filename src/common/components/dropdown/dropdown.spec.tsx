import '@testing-library/jest-dom/extend-expect';
import { h, VNode } from 'preact';
import { render } from '@testing-library/preact';
import { Dropdown, DropdownOption } from './dropdown.component';
import { Text } from '../../../modules/localisation/components/text';

const data: DropdownOption[] = [
    {
        id: 0,
        label: 'Mateusz',
        value: { firstName: 'Mateusz', lastName: 'Kosowski' },
    },
    {
        id: 1,
        label: 'Janusz',
        value: { firstName: 'Janusz', lastName: 'Pindera' },
    },
    {
        id: 2,
        label: 'Stanisław',
        value: { firstName: 'Stanisław', lastName: 'Mendoza' },
    },
];

describe('Dropdown', () => {
    it('should render correctly', () => {
        const { asFragment } = render(
            <Dropdown
                renderSelectedOption={(selectedOption): VNode => {
                    return <Text>{selectedOption.label}</Text>;
                }}
                renderOptionList={(OptionComponent, onOptionClick): VNode[] => {
                    return data.map(user => (
                        <OptionComponent
                            key={user.id}
                            onClick={(): void => onOptionClick(user)}
                        >
                            {user.label}
                        </OptionComponent>
                    ));
                }}
            />
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
