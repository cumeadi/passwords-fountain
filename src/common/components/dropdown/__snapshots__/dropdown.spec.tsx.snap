// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should render correctly 1`] = `
<DocumentFragment>
  <section
    class="go62670245"
    tabindex="0"
  >
    <div
      class="go3942427238"
    >
      Choose
      <span
        class="go4075838652"
      />
    </div>
    <div
      class="go129576329"
    />
  </section>
</DocumentFragment>
`;
